Shared (App)/ViewModels/TransactionsViewModel.swift

```diff
@@ -103,7 +123,32 @@ final class TransactionsListViewModel: ObservableObject {
     private func canLoadNextPage(atTransaction transaction: TransactionGroup) -> Bool {
         guard let index: Int = transactions.firstIndex(of: transaction) else { return false }
         let reachedThreshold = Double(index) / Double(transactions.count) > 0.7
-        return !isFetching && reachedThreshold // TODO: check if end has been reached
+        return !isFetching && reachedThreshold
+    }
+    
+    @MainActor
+    private func fetchContracts(fromTxs txs: [TransactionGroup]) async {
+        var contracts = [Contract]()
+        await withTaskGroup(of: Contract?.self) { [weak self] group in
+            guard let self = self else { return }
+            for tx in txs {
+                group.addTask {
+                    guard let contractAddress = tx.transactions.first?.to,
+                          self.contracts[tx.toAddress] == nil else { return nil }
+                    // TODO: Fetch tx input and decode with ABI fetched here
+                    // Discuss if we want to fetch tx input and contract details on list or on tx detail screen
+                    // If we want to show all details on the list, performance may be laggy because of the numerous fetches
+                    return try? await self.contractService.fetchContractDetails(forAddress: contractAddress)
+                }
+                for await contract in group {
+                    guard let contract = contract else { return }
+                    contracts.append(contract)
+                }
+            }
+        }
+        for contract in contracts {
+            self.contracts[contract.address] = contract
```

Comment:
Using task groups (similar to dispatch groups) in order to make it faster to fetch contract details (once we fetch from etherscan). Ran into data race issues which is why I'm doing multiple loops here. Can probably be improved.